text/0784-no-globals.md

```diff
@@ -0,0 +1,494 @@
+implementation details, which are not prescribed in this RFC but hopefully useful for any future implementor, so feel free
+to skip over this section:
+
+---
+
+<details>
+<summary>Compilation changes</summary>
+
+A typical build of `vendor.js` will look basically like this:
+
+<details>
+<summary>vendor.js</summary>
+
+```js
+// EmberENV code, ommited here, see following chapter
+
+var loader, define, requireModule, require, requirejs;
+
+(function(global) {
+  
+  // loader.js IIFE defing the globals above
+  // see https://github.com/ember-cli/loader.js/blob/master/lib/loader/loader.js
+
+})(this);
+
+(function() {
+  var define, require;
+  
+  (function () {
+    // Ember's own internal loader code
+  
+    // ...
+    // globalObj = globalThis w/ polyfill
+    if (typeof globalObj.define === 'function' && typeof globalObj.require === 'function') {
+      define = globalObj.define;
+      require = globalObj.require;
+  
+      return;
+    }
+    
+    // Following are Ember's own (internal) implementation of define and require
+    // However due to the early return above and the fact that loader.js always prepends its code, thus assigns
+    // the define and require globals, this code is never reached! (when the loader.js addon is present!)
+    
+  })();
+  define("@ember/-internals/bootstrap/index", /*...*/);
+  // ... 
+  // all of Ember's own modules
+  // ...
+  require('@ember/-internals/bootstrap')
+})();
+
+define('@ember-data/adapter/-private', /* ... */);
+// ... 
+// all modules from Ember addons
+// ...
+```
+</details>
+
+And this is what a typical compilation of `app.js` will look like:
+
+<details>
+<summary>app.js</summary>
+
+```js
+define("my-app/app", /* ... */);
+// ... 
+// define() for all app modules
+// ...
+```
+</details>
+
+As you can see it relies on a globally set `define`, that has been declared in the first line of `vendor.js`, and [set
+by the code of `loader.js`](https://github.com/ember-cli/loader.js/blob/ed8c85a713b0dac23bc974e0850863412304c1b1/lib/loader/loader.js#L46).
+
+As to how to prevent the global usage while retaining compatibility of any code that uses those functions *as globals*,
+we can make the build system emit the existing code wrapped in a [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE),
+that avoids polluting the global namespace, while still any calls of `define()` or `require()` will see the same functions,
+they just do not come from the global namespace anymore.
+
+This works for `vendor.js`, where these functions are defined, but not for `app.js` which needs access to the *same* functions
+as defined in `vendor.js`. That's why we still need *some* usage of globals, but as outlined in the previous chapter, we
+must make sure they are scoped to the current app, so they pose no risk of conflicts.
+
+So for `vendor.js` this is what the basic change could look like:
+
+
+<details>
+<summary>vendor.js</summary>
+
+```diff
++ (function(){
```

Comment:
Also note that I mentioned under "Alternatives" that we could also drop all support of user-land code calling the AMD functions. Which would be more invasive, breaking some addon's code. Still no breaking change, as that is opt-in only.

Then indeed we don't need that IIFE nor any addon code transformations...

Reply:
> But wouldn't that break any addon code that still uses e.g. require()?

Yes 😈 , but that's the point of creating an optional feature anyways (it's opt-in, we are allowed to limit semantics).

---

> The idea here was to prevent leaking those into the global namespace, while still allowing all existing addon code that uses these things in the wild (most notably ember-resolver) to still work. That is as long as that addon code is processed through the normal Ember CLI build process, i.e. it lands within the IIFE. When you call require from somewhere else like an external <script>, that would cease to work, hence the optional feature to allow this "intimate API" to break in such rare cases.

Ya, I think we should provide a global that is app scoped (e.g. `globalThis._appNameGlobals.require` and `globalThis._appNameGlobals.define`) instead of continuing to use `define`/`require` unscoped. This would require some tweaks to the babel setup (e.g. we'd need a custom version of https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-modules-amd/src/index.ts or send a PR to allow the `define`/`require` hard coded constants to change).

---

> Note that I am referring to addons directly calling require() etc., like [here](https://github.com/miragejs/ember-cli-mirage/blob/d99571d9010011976b757f40cca8450c71dcb070/packages/ember-cli-mirage/addon/utils/read-modules.js#L62), and not regular ESM imports that happen to be transformed to define/require during the build.

I actually think this case is **not** a good example. Specifically because that is not using `require` global, it is using `import require from 'require';`:


https://github.com/miragejs/ember-cli-mirage/blob/d99571d9010011976b757f40cca8450c71dcb070/packages/ember-cli-mirage/addon/utils/read-modules.js#L8

^ pattern should still continue to work just fine (because whatever loader happens to be running, will respond to that import with the correctly scoped `require`).

Reply:
> Yes 😈 , but that's the point of creating an optional feature anyways (it's opt-in, we are allowed to limit semantics).

Yes, which is basically what I say here, right? 👇

> Also note that I mentioned under "Alternatives" that we could also drop all support of user-land code calling the AMD functions. Which would be more invasive, breaking some addon's code. Still no breaking change, as that is opt-in only.

At least when they are called as globals. Which the current form of the RFC still tries to support...

So to get this 100% clear: you are suggesting that we do this alternative approach, i.e. stop supporting the global use of `require()` & Co as globals all together, right? 

This does mean that we need to "fix" a lot more of the core addons, like `ember-resolver`, which AFAICT does use `require()` as a global [here](https://github.com/ember-cli/ember-resolver/blob/master/addon/resolvers/classic/index.js)

> I actually think this case is not a good example. Specifically because that is not using require global

Oh, you are right, I missed that! It's not so easy to search for examples of addons doing `require()` using emberobserver's code search, given that you get a bazillion hits for legit use of `require()` in node-land... 🙈

>  pattern should still continue to work just fine (because whatever loader happens to be running, will respond to that import with the correctly scoped require)

While that is possible, I wonder if we *actually* want that. Probably out of scope for this RFC, but shouldn't we eventually try to get rid of any usage of loader.js stuff at all? 

Like in a future where we have Polaris and Embroider by default, shouldn't we tell users that the only way to import stuff is via ESM imports (and `@embroider/macros`)? So that we can *eventually* get to a point where we could get rid of any specific runtime module system like loader.js, when we have just ESM imports at stage 3, then throw that at webpack, which then emits whatever it emits. 

So maybe turning on `no-globals` is actually a good moment to make *all* of `require()` stop working for unprivileged code?